File: LeetCode/remote/maximal_square.cs
// https://leetcode.com/problems/maximal-square/
//
//  Given a 2D binary matrix filled with 0's and 1's, find the largest square containing all 1's and return its area.
//
//  For example, given the following matrix:
//
//  1 0 1 0 0
//  1 0 1 1 1
//  1 1 1 1 1
//  1 0 0 1 0
//
//  Return 4. 
//
// 
// Submission Details
// 67 / 67 test cases passed.
//  Status: Accepted
//  Runtime: 156 ms
//      
//      Submitted: 0 minutes ago
//

using System;

public class Solution {
    public int MaximalSquare(char[,] matrix) {
        var mins = new int[matrix.GetLength(0), matrix.GetLength(1)];
        int max = 0;
        for (int i = 0; i < matrix.GetLength(0); i++)
        {
            mins[i, 0] = matrix[i, 0] == '1' ? 1 : 0;
            max = Math.Max(max, mins[i, 0]);
        }

        for (int i = 0; i < matrix.GetLength(1); i++)
        {
            mins[0, i] = matrix[0, i] == '1' ? 1 : 0;
            max = Math.Max(max, mins[0, i]);
        }
        
        for (int i = 1; i < matrix.GetLength(0); i++)
        {
            for (int j = 1; j < matrix.GetLength(1); j++)
            {
                if (matrix[i, j] == '1')
                {
                    mins[i, j] = Math.Min(mins[i - 1, j - 1], Math.Min(mins[i - 1, j], mins[i, j - 1])) + 1;
                    max = Math.Max(max, mins[i, j]);
                }
            }
        }

        return max * max;
    }

    //  https://leetcode.com/submissions/detail/62912583/
    //
    //  Submission Details
    //  67 / 67 test cases passed.
    //      Status: Accepted
    //      Runtime: 164 ms
    //          
    //          Submitted: 0 minutes ago
    public int MaximalSquare(char[,] matrix) {
        var max = 0;
        for (var i = 0; i < matrix.GetLength(0); i++) {
            for (var j = 0; j < matrix.GetLength(1); j++) {
                if (matrix[i, j] == '0') {
                    continue;
                }
                
                var up = i == 0 ? 0 : matrix[i - 1, j] - '0';
                var left = j == 0 ? 0 : matrix[i, j - 1] - '0';
                var upLeft = i == 0 || j == 0 ? 0 : matrix[i - 1, j - 1] - '0';
                matrix[i, j] = (char) ('0' + 1 + Math.Min(upLeft, Math.Min(up, left)));
                max = Math.Max(max, (int) Math.Pow(matrix[i, j] - '0', 2));
            }
        }
        
        return max;
    }

    static void Main()
    {
        Console.WriteLine(new Solution().MaximalSquare(new char[,] {
                                                        {'1', '0', '1', '0', '0'},
                                                        {'1', '0', '1', '1', '1'},
                                                        {'1', '1', '1', '1', '1'},
                                                        {'1', '0', '0', '1', '0'}
                                                      }));
    }
}

File: LeetCode/remote/maximal_square.rb
#   https://leetcode.com/problems/maximal-square/
#
#    Given a 2D binary matrix filled with 0's and 1's, find the largest square containing all 1's and return its area.
#   https://leetcode.com/submissions/detail/53964916/
#
#   Submission Details
#   67 / 67 test cases passed.
#       Status: Accepted
#       Runtime: 300 ms
#           
#           Submitted: 0 minutes ago

def maximal_square(matrix)
    return 0 if matrix.nil? || matrix.empty?
    n = matrix.size
    m = matrix[0].size
    dp = Array.new(n) {
            Array.new(m) {
                0
            }
         }
    
    (0...n).to_a
    .product((0...m).to_a)
    .select { |i, j| matrix[i][j] != '0' }
    .map { |i, j|
        l = j == 0 ? 0 : dp[i][j - 1]
        u = i == 0 ? 0 : dp[i - 1][j]
        ul = i == 0 && j == 0 ? 0 : dp[i - 1][j - 1]
        dp[i][j] = [l, u, ul].min + 1
        dp[i][j] * dp[i][j]
    }
    .max || 0
end
